megamek/src/megamek/common/TargetRoll.java

```diff
@@ -23,75 +29,72 @@
-    /*
-     * The CHECK_FALSE value is returned when a function that normally would
-     * return a target roll number determines that the roll wasn't needed after
-     * all.
-     */
-    public static final int CHECK_FALSE = Integer.MIN_VALUE + 1;
 
-    private List<TargetRollModifier> modifiers = new ArrayList<>();
+    /** The CHECK_FALSE value is returned when a function that normally would return a target roll number
+     determines that the roll wasn't needed after all. */
+    public static final int CHECK_FALSE = Integer.MIN_VALUE + 1;
 
+    /**
+     * This list of roll modifiers. Always call recalculate() after modifying it. This is clearly an unsafe
+     * way to implement it. It *may be* done like that for performance reasons for Princess.
+     */
+    private final List<TargetRollModifier> modifiers = new ArrayList<>();
     private int total;
 
     /**
-     * Creates new TargetRoll
+     * Creates a new, empty TargetRoll. This is by itself not yet useful.
      */
-    public TargetRoll() {
-
-    }
+    public TargetRoll() { }
 
     /**
      * Creates a new TargetRoll with a base value and desc
      */
     public TargetRoll(int value, String desc) {
-        addModifier(value, desc);
+        this(new TargetRollModifier(value, desc));
     }
 
     /**
-     * Creates a new TargetRoll with a base value and desc, which is possibly
-     * not cumulative
-     *
-     * @param value
-     * @param desc
-     * @param cumulative
+     * Creates a new TargetRoll with a base value and desc, which is possibly not cumulative
      */
     public TargetRoll(int value, String desc, boolean cumulative) {
-        addModifier(value, desc, cumulative);
+        this(new TargetRollModifier(value, desc, cumulative));
+    }
+
+    public TargetRoll(TargetRollModifier baseValue) {
+        addModifier(baseValue);
     }
 
     /**
-     * Returns the total value of all modifiers
+     * @return The sum of all modifiers. When a finalizer is present, the value equals that finalizer.
      */
     public int getValue() {
         return total;
     }
 
     /**
-     * Returns the total value of all modifiers
+     * @return The total value of all modifiers as a text.
      */
     public String getValueAsString() {
-        switch (total) {
-            case IMPOSSIBLE:
-                return "Impossible";
-            case AUTOMATIC_FAIL:
-                return "Automatic Failure";
-            case AUTOMATIC_SUCCESS:
-                return "Automatic Success";
-            case CHECK_FALSE:
-                return "Did not need to roll";
-            default:
-                return Integer.toString(total);
-        }
+        return switch (total) {
+            case IMPOSSIBLE -> "Impossible";
+            case AUTOMATIC_FAIL -> "Automatic Failure";
+            case AUTOMATIC_SUCCESS -> "Automatic Success";
+            case CHECK_FALSE -> "Did not need to roll";
+            default -> Integer.toString(total);
+        };
```

Comment:
So fresh and so clean!

Reply:
yeah :) although strictly speaking this should be localized...